src/home/homework-blob.c

```diff
@@ -0,0 +1,236 @@
+                const char *src_name,
+                int dest_dfd,
+                const char *dest_name,
+                Hashmap *manifest,
+                mode_t mode,
+                uid_t uid) {
+        _cleanup_free_ char *fn = NULL;
+        _cleanup_close_ int src_fd = -EBADF, dest_fd = -EBADF;
+        struct blob_copy_data userdata;
+        int r;
+
+        assert(src_dfd >= 0);
+        assert(src_name);
+        assert(dest_dfd >= 0);
+        assert(dest_name);
+
+        src_fd = openat(src_dfd, src_name, O_RDONLY|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
+        if (src_fd < 0) {
+                if (errno == ENOENT)
+                        return 0;
+                return log_debug_errno(errno, "Failed to open src blob dir: %m");
+        }
+
+        r = tempfn_random(dest_name, NULL, &fn);
+        if (r < 0)
+                return r;
+
+        dest_fd = open_mkdir_at(dest_dfd, fn, O_EXCL|O_CLOEXEC, mode);
+        if (dest_fd < 0)
+                return log_debug_errno(dest_fd, "Failed to create/open dest blob dir: %m");
+
+        /* Note: We do it this way instead of just using FOREACH_DIRENT so that we
+         * walk the dirents in alphabetical order and thus behave deterministically
+         * w.r.t. what happens if the dir hits its size quota */
+        userdata = (struct blob_copy_data) {
+                .dest_dfd = dest_fd,
+                .total_size = 0,
+                .uid = uid,
+                .manifest = manifest,
+        };
+        r = recurse_dir_at(src_fd, ".", STATX_TYPE|STATX_SIZE, UINT_MAX, RECURSE_DIR_SORT,
+                           blob_copy_callback, &userdata);
+        if (r < 0) {
+                r = log_debug_errno(r, "Failed to fill blob dir: %m");
+                goto fail;
+        }
+
+        if (fchown(dest_fd, uid, uid) < 0) {
+                r = log_debug_errno(errno, "Failed to chown dest blob: %m");
+                goto fail;
+        }
+
+        r = install_file(dest_dfd, fn, dest_dfd, dest_name, INSTALL_REPLACE);
+        if (r < 0) {
+                r = log_debug_errno(errno, "Failed to move dest blob into place: %m");
+                goto fail;
+        }
+
+        return 0;
+
+fail:
+        (void) rm_rf_at(dest_dfd, fn, REMOVE_ROOT|REMOVE_PHYSICAL|REMOVE_MISSING_OK);
+        return r;
+}
+
+int home_reconcile_blob_dirs(UserRecord *h, int root_fd, int reconciled) {
+        _cleanup_close_ int sys_base_dfd = -EBADF;
+        int r;
+
+        assert(h);
+        assert(root_fd >= 0);
+        assert(reconciled >= 0);
+
+        if (reconciled == USER_RECONCILE_IDENTICAL)
+                return 0;
+
+        sys_base_dfd = open(home_system_blob_dir(), O_RDONLY|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
+        if (sys_base_dfd < 0)
+                return log_error_errno(errno, "Failed to open system blob dir: %m");
+
+        if (reconciled == USER_RECONCILE_HOST_WON) {
+                r = replace_blob_at(sys_base_dfd, h->user_name, root_fd, ".identity-bulk",
```

Comment:
bulk? blob?

Reply:
gah I'm still not used to it being a blob dir rather than bulk dir, so fingers automatically type bulk. I thought I got rid of all the instances of this, but apparently I missed the most important one :face_exhaling: